File: androidApp/src/main/java/org/blackcandy/android/bridge/ThemeComponent.kt
package org.blackcandy.android.bridge

import dev.hotwire.core.bridge.BridgeComponent
import dev.hotwire.core.bridge.BridgeDelegate
import dev.hotwire.core.bridge.Message
import dev.hotwire.navigation.destinations.HotwireDestination
import kotlinx.serialization.Serializable
import org.blackcandy.android.fragments.web.WebFragment
import org.blackcandy.shared.utils.Theme
import org.blackcandy.shared.viewmodels.WebViewModel

class ThemeComponent(
    name: String,
    private val delegate: BridgeDelegate<HotwireDestination>,
) : BridgeComponent<HotwireDestination>(name, delegate) {
    private val viewModel: WebViewModel
        get() {
            val fragment = delegate.destination.fragment as WebFragment
            return fragment.viewModel
        }

    override fun onReceive(message: Message) {
        when (message.event) {
            "initialize" -> handleInitializeEvent(message)
        }
    }

    private fun handleInitializeEvent(message: Message) {
        val data = message.data<ThemeData>() ?: return
        val themeValue = Theme.values().find { it.name == data.theme.uppercase() } ?: return

        viewModel.updateTheme(themeValue)
    }

    @Serializable
    data class ThemeData(
        val theme: String,
    )
}

File: iosApp/iosApp/Bridge/ThemeComponent.swift
import HotwireNative
import sharedKit

class ThemeComponent: BridgeComponent {
    override class var name: String { "theme" }

    private var viewModel: WebViewModel? {
        let viewController = delegate?.destination as? WebViewController
        return viewController?.viewModel
    }

    override func onReceive(message: Message) {
        switch message.event {
        case "initialize":
            handleInitializeEvent(message)
        default:
            break
        }
    }

    private func handleInitializeEvent(_ message: Message) {
        guard let data: ThemeData = message.data() else { return }
        guard let themeValue = Theme.allCases.first(where: { $0.name == data.theme.uppercased() }) else { return }

        viewModel?.updateTheme(theme: themeValue)
    }
}

extension ThemeComponent {
    struct ThemeData: Decodable {
        let theme: String
    }
}
